Sources/MessagingPush/MessagingPush.swift

```diff
@@ -7,43 +7,70 @@ import Foundation
   */
 public class MessagingPush: ModuleTopLevelObject<MessagingPushInstance>, MessagingPushInstance {
     @Atomic public private(set) static var shared = MessagingPush()
+    private static let moduleName = "MessagingPush"
+
     private var globalDataStore: GlobalDataStore
 
     // testing constructor
     init(implementation: MessagingPushInstance?, globalDataStore: GlobalDataStore) {
         self.globalDataStore = globalDataStore
-        super.init(implementation: implementation)
+        super.init(moduleName: MessagingPush.moduleName, implementation: implementation)
     }
 
     // singleton constructor
-    override private init() {
+    private init() {
         self.globalDataStore = CioGlobalDataStore.getInstance()
-        super.init()
+        super.init(moduleName: MessagingPush.moduleName)
     }
 
     // for testing
     static func resetSharedInstance() {
         shared = MessagingPush()
     }
 
-    // At this time, we do not require `MessagingPush.initialize()` to be called to make the SDK work. There is
-    // currently no module initialization to perform.
-    public static func initialize() {
-        MessagingPush.shared.inititlizeModule()
+    /**
+     Initialize the shared `instance` of `MessagingPush`.
+     Call this function when your app launches, before using `MessagingPush.shared`.
+     */
+    @discardableResult
+    public static func initialize(
+        configure configureHandler: ((inout MessagingPushConfigOptions) -> Void)? = nil
+    ) -> MessagingPushInstance {
+        var moduleConfig = MessagingPushConfigOptions.Factory.create()
+
+        if let configureHandler = configureHandler {
+            configureHandler(&moduleConfig)
+        }
+
+        shared.initializeModule(moduleConfig: moduleConfig)
+        return shared
```

Comment:
Similar to current initialization, push module configuration can be updated using any of the following option

1. Using optional initialization

```swift
MessagingPush.initialize { config in
    config.autoTrackPushEvents = true
}
```

2. Using `configure` method

```swift
MessagingPush.shared.configure { config in
    config.autoTrackPushEvents = true
}
```

Reply:
I like the idea of creating a separate `configure()` function to allow the SDK to change it's config at runtime instead of config being required during SDK initialization. 

I have thought about how we could add this type of feature to our SDK but realized that it could cause scope creep or risk of runtime errors. 

Here is a scenario to describe this risk. 

1. `MessagingPush.initialize` gets called and sets the config to: 
```
autoTrackPushMetrics = true
```

2. Some code in the push module get executed. So, some dependencies in the DI graph get constructed. Here is an example constructor that you would see in the iOS codebase: 

```swift
class PushClickHandler {

  let autoTrackPushMetrics: Bool

  init(moduleConfig: ModulePushConfigOptions) {
    autoTrackPushMetrics = moduleConfig.autoTrackPushMetrics
  }
}
```

3. `MessagingPush.shared.configure` gets called and changes `autoTrackPushMetrics = false`. 

4. The existing `PushClickHandler` instance could still be in-memory and it contains the old config `autoTrackPushMetrics` value. Putting the SDK into a state where some dependencies are out-of-date. 

---

There are ways to avoid this scenario from happening. The push module is small in size so the changes of this scenario happening are much smaller then other modules such as Track or CDP. 

Even if we do not see this problem in the push module today, this is a easy risk to fall into. I think that the way the codebase is today, it's far too easy to accidentally write code that puts the SDK into this state. Therefore, I am nervous to add this feature at this time. 

If adding this feature is required to make the SDK function, a possible workaround is to only allow each module to be configured one time. No matter what function they call, `initialize` or `config`, we set module configuration one time and ignore future requests to change the config. 

I would like to add this feature to all of our modules. I would feel most confident if it was done when we could reduce the risk by making some refactors to the codebase first.

Reply:
Thanks for sharing the use case, Levi. I agree that even if we address all current instances where the module config is cached, there's still a high chance of it being cached and misused in the future. I've [moved this discussion to Slack](https://customerio.slack.com/archives/C05UY5J4VLK/p1701251917423449) for detailed explanation of what I had in mind when adding this. Let’s discuss further there and decide on our next action items.